File: packages/apple/Sources/Composable/Chain.swift
import Foundation

// MARK: - Chain Protocol

/// Core protocol for composable AI processing pipelines.
///
/// A Chain takes a `ChainInput` and produces a `ChainOutput`.
/// Chains can be composed sequentially, in parallel, or conditionally.
///
/// ```swift
/// let pipeline = SequentialChain(chains: [summarizeChain, translateChain])
/// let result = try await pipeline.invoke(ChainInput(text: article))
/// ```
@available(iOS 15.0, macOS 14.0, *)
public protocol Chain: Sendable {
    /// Human-readable name for logging and debugging
    var name: String { get }

    /// Execute this chain with the given input
    func invoke(_ input: ChainInput) async throws -> ChainOutput
}

// MARK: - Sequential Chain

/// Composes chains to run sequentially, passing output → input
@available(iOS 15.0, macOS 14.0, *)
public struct SequentialChain: Chain {
    public let name: String
    private let chains: [any Chain]

    public init(name: String = "SequentialChain", chains: [any Chain]) {
        self.name = name
        self.chains = chains
    }

    public func invoke(_ input: ChainInput) async throws -> ChainOutput {
        var current = input
        var lastOutput: ChainOutput?

        for chain in chains {
            let output = try await chain.invoke(current)
            current = ChainInput(text: output.text, metadata: output.metadata)
            lastOutput = output
        }

        guard let result = lastOutput else {
            throw LocanaraError.executionFailed("SequentialChain has no chains")
        }
        return result
    }
}

// MARK: - Parallel Chain

/// Runs chains in parallel and collects all results
@available(iOS 15.0, macOS 14.0, *)
public struct ParallelChain: Chain {
    public let name: String
    private let chains: [any Chain]

    public init(name: String = "ParallelChain", chains: [any Chain]) {
        self.name = name
        self.chains = chains
    }

    public func invoke(_ input: ChainInput) async throws -> ChainOutput {
        let results = try await withThrowingTaskGroup(of: ChainOutput.self) { group in
            for chain in chains {
                group.addTask {
                    try await chain.invoke(input)
                }
            }

            var outputs: [ChainOutput] = []
            for try await result in group {
                outputs.append(result)
            }
            return outputs
        }

        let combinedText = results.map(\.text).joined(separator: "\n---\n")
        var combinedMetadata = input.metadata
        for result in results {
            combinedMetadata.merge(result.metadata) { _, new in new }
        }

        return ChainOutput(
            value: results,
            text: combinedText,
            metadata: combinedMetadata
        )
    }
}

// MARK: - Conditional Chain

/// Routes to different chains based on a condition function
@available(iOS 15.0, macOS 14.0, *)
public struct ConditionalChain: Chain {
    public let name: String
    private let condition: @Sendable (ChainInput) -> String
    private let branches: [String: any Chain]
    private let defaultChain: (any Chain)?

    public init(
        name: String = "ConditionalChain",
        condition: @escaping @Sendable (ChainInput) -> String,
        branches: [String: any Chain],
        defaultChain: (any Chain)? = nil
    ) {
        self.name = name
        self.condition = condition
        self.branches = branches
        self.defaultChain = defaultChain
    }

    public func invoke(_ input: ChainInput) async throws -> ChainOutput {
        let branchKey = condition(input)
        guard let chain = branches[branchKey] ?? defaultChain else {
            throw LocanaraError.executionFailed(
                "No branch found for key '\(branchKey)' in ConditionalChain"
            )
        }
        return try await chain.invoke(input)
    }
}

// MARK: - Model Chain

/// A simple chain that sends input to a model and returns the response.
/// This is the basic building block for custom chains.
@available(iOS 15.0, macOS 14.0, *)
public struct ModelChain: Chain {
    public let name: String
    private let model: any LocanaraModel
    private let promptTemplate: PromptTemplate?
    private let config: GenerationConfig?

    public init(
        name: String = "ModelChain",
        model: (any LocanaraModel)? = nil,
        promptTemplate: PromptTemplate? = nil,
        config: GenerationConfig? = nil
    ) {
        self.name = name
        self.model = model ?? LocanaraDefaults.model
        self.promptTemplate = promptTemplate
        self.config = config
    }

    public func invoke(_ input: ChainInput) async throws -> ChainOutput {
        let prompt: String
        if let template = promptTemplate {
            var values = input.metadata
            values["text"] = input.text
            prompt = try template.format(values)
        } else {
            prompt = input.text
        }

        let response = try await model.generate(prompt: prompt, config: config)

        return ChainOutput(
            value: response.text,
            text: response.text,
            metadata: input.metadata,
            processingTimeMs: response.processingTimeMs
        )
    }
}

File: packages/android/locanara/src/main/kotlin/com/locanara/composable/Chain.kt
package com.locanara.composable

import com.locanara.core.ChainInput
import com.locanara.core.ChainOutput
import com.locanara.core.GenerationConfig
import com.locanara.core.LocanaraDefaults
import com.locanara.core.LocanaraModel
import com.locanara.core.PromptTemplate
import kotlinx.coroutines.async
import kotlinx.coroutines.awaitAll
import kotlinx.coroutines.coroutineScope

/**
 * Core interface for composable AI processing pipelines.
 *
 * ```kotlin
 * val pipeline = SequentialChain(chains = listOf(summarizeChain, translateChain))
 * val result = pipeline.invoke(ChainInput(text = article))
 * ```
 */
interface Chain {
    /** Human-readable name for logging and debugging */
    val name: String

    /** Execute this chain with the given input */
    suspend fun invoke(input: ChainInput): ChainOutput
}

/**
 * Composes chains to run sequentially, passing output → input
 */
class SequentialChain(
    override val name: String = "SequentialChain",
    private val chains: List<Chain>
) : Chain {
    override suspend fun invoke(input: ChainInput): ChainOutput {
        var current = input
        var lastOutput: ChainOutput? = null

        for (chain in chains) {
            val output = chain.invoke(current)
            current = ChainInput(text = output.text, metadata = output.metadata)
            lastOutput = output
        }

        return lastOutput ?: throw IllegalStateException("SequentialChain has no chains")
    }
}

/**
 * Runs chains in parallel and collects results
 */
class ParallelChain(
    override val name: String = "ParallelChain",
    private val chains: List<Chain>
) : Chain {
    override suspend fun invoke(input: ChainInput): ChainOutput = coroutineScope {
        val results = chains.map { chain ->
            async { chain.invoke(input) }
        }.awaitAll()

        val combinedText = results.joinToString("\n---\n") { it.text }
        val combinedMetadata = input.metadata.toMutableMap()
        results.forEach { combinedMetadata.putAll(it.metadata) }

        ChainOutput(
            value = results,
            text = combinedText,
            metadata = combinedMetadata
        )
    }
}

/**
 * Routes to different chains based on a condition
 */
class ConditionalChain(
    override val name: String = "ConditionalChain",
    private val condition: (ChainInput) -> String,
    private val branches: Map<String, Chain>,
    private val defaultChain: Chain? = null
) : Chain {
    override suspend fun invoke(input: ChainInput): ChainOutput {
        val branchKey = condition(input)
        val chain = branches[branchKey] ?: defaultChain
            ?: throw IllegalStateException("No branch for key '$branchKey'")
        return chain.invoke(input)
    }
}

/**
 * A simple chain that sends input to a model and returns the response.
 */
class ModelChain(
    override val name: String = "ModelChain",
    private val model: LocanaraModel = LocanaraDefaults.model,
    private val promptTemplate: PromptTemplate? = null,
    private val config: GenerationConfig? = null
) : Chain {
    override suspend fun invoke(input: ChainInput): ChainOutput {
        val prompt: String = if (promptTemplate != null) {
            val values = input.metadata.toMutableMap()
            values["text"] = input.text
            promptTemplate.format(values)
        } else {
            input.text
        }

        val response = model.generate(prompt, config)

        return ChainOutput(
            value = response.text,
            text = response.text,
            metadata = input.metadata,
            processingTimeMs = response.processingTimeMs
        )
    }
}
